### PR TITLE
Upgrade to latest version of Alloy and port Anvil tests

### LINKED ISSUE #7449: anvil test migration to alloy
### Component

Anvil

### Describe the feature you would like

anvil e2e tests are using ethers-providers, which can be fully migrated to foundry now:

https://github.com/foundry-rs/foundry/blob/0026488754512acf0fd902a0d2c90cf8a09367b0/crates/anvil/tests/it/main.rs#L1-L1

these can be done iteratively, one file at a time so that we could work in tandem on this

the setup should be very similar to alloy-examples

cc @zerosnacks @yash-atreya 

~~likely blocked until @klkvr cast migration pr lands that updates alloy~~

Unblocked

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Builds on top of #7619 
Closes #7449 

## Solution

Upgrades Foundry to latest version of Alloy

Updates the following examples to use Alloy

- [x] abi.rs
- [x] anvil.rs
- [x] anvil_api.rs
- [x] api.rs
- [x] fork.rs
- [x] ganache.rs
- [x] gas.rs
- [x] genesis.rs
- [x] geth.rs
- [x] ipc.rs
- [x] logs.rs
- [x] optimism.rs
- [x] otterscan.rs
- [x] proof.rs
- [x] pubsub.rs
- [x] revert.rs
- [x] sign.rs
- [x] state.rs
- [x] traces.rs
- [x] transaction.rs
- [x] txpool.rs
- [x] utils.rs
- [x] wsapi.rs

**Breaking changes**:

- Remove special case handling of `Polygon` and `Polygon Mumbai` EIP-1559 fees handling of GasOracle middleware

**Notes**:

- The final migration of `cast bind` and `forge bind` is planned to be handled in a standalone ticket given its possible complexity and the diff has already grown massively
- Given that the port of Alloy has largely already occurred the odd cases of failing tests are largely descriptive of current issues (and flagged as such), not new issues introduced by this ticket.

- Few tests have been migrated but fail due to issues:

1. ~https://github.com/alloy-rs/alloy/issues/389~
~Tests failing due to `test_fork_snapshotting_blocks`,  `test_fork_timestamp`, `test_fork_reset_moonbeam`, `can_call_ots_get_transaction_error`, `can_replace_transaction`.~

2. RpcError in `watch_full_pending_transactions`
```bash
ErrorPayload { code: -32602, message: "invalid type: boolean `true`, expected
// unit", data: None } originating from watch_full_pending_transactions
```
~3. Test failing: `can_stream_pending_transactions`: Pubsub `channel_size` issue in `test_sub_new_heads_fast`.~
4. Optimism `test_deposits_not_supported_if_optimism_disabled`, likely the other Optimism tests aren't working as expected
5. Test failing: `can_get_block_number_ipc` and `test_sub_new_heads_ipc` fail on Windows

### CODE REVIEW

**Comment by zerosnacks** [comment]
From @yash-atreya 

- `ethers` has been removed from all anvil tests except `optimism` tests (@zerosnacks is on it).

- Few tests have been migrated but fail due to issues:

1. https://github.com/alloy-rs/alloy/issues/389 
Tests failing due to `test_fork_snapshotting_blocks`,  `test_fork_timestamp`, `test_fork_reset_moonbeam`, `can_call_ots_get_transaction_error`, `can_replace_transaction`.

2. RpcError in `watch_full_pending_transactions`
```bash
ErrorPayload { code: -32602, message: "invalid type: boolean `true`, expected
// unit", data: None } originating from watch_full_pending_transactions
```
Test failing: `can_stream_pending_transactions`

3. Pubsub `channel_size` issue in `test_sub_new_heads_fast`.

Once, `optimism` tests have been migrated, we should proceed to remove `ethers` completely from the codebase by tackling the following in no particular order: 

1. Migrate `cast bind` and `forge bind`
2. Doc tests in `cast/lib.rs`
3. Remove `ethers` provider from `foundry_common`. This can be done as soon as `optimisim` is migrated.
4. `RuntimeClient` in `foundry_common`
5. `RuntimeTransport` in `foundry_common`
6. forge cli tests 
7. Finally remove the types `ToAlloy` and `ToEthers` from `foundry_common`

**Comment by zerosnacks** [comment]
@yash-atreya I've marked this as ready for review now that all tests are running. I think it is best if the final migration of `cast bind` and `forge bind` is handled in a standalone ticket given its possible complexity and the diff is already massive.

File: crates/anvil/tests/it/pubsub.rs [hunk 1]
```diff
@@ -1,275 +1,278 @@
...
+
+    let num = 1000u64;
+
+    let mut block_numbers = Vec::new();
+    for _ in 0..num {
+        let mine = api.mine_one();
+        let block = async {
+            let block = blocks.next().await.unwrap();
+            block.header.number.unwrap()
+        };
```
Comment (mattsse) [inline crates/anvil/tests/it/pubsub.rs:269]: nit: the async scope is not needed here and can just be extracted and put after the mine.await

File: crates/anvil/tests/it/transaction.rs [hunk 2]
```diff
@@ -1,212 +1,228 @@
...
     assert_eq!(2, block.transactions.len());
-    assert_eq!(vec![tx.transaction_hash, higher_tx.transaction_hash], block.transactions);
+    assert_eq!(
+        BlockTransactions::Hashes(vec![tx.transaction_hash, higher_tx.transaction_hash]),
+        block.transactions
+    );
 }
 
+// TODO: Revisit: https://github.com/alloy-rs/alloy/issues/389 didn't fix the issue. Remove ignore.
+#[ignore]
```
Comment (mattsse) [inline crates/anvil/tests/it/transaction.rs:141]: this shoudl not be ignored

File: crates/common/src/provider/mod.rs [hunk 3]
```diff
@@ -1,7 +1,331 @@
...
+        JoinFill<JoinFill<JoinFill<Identity, GasFiller>, NonceFiller>, ChainIdFiller>,
+        SignerFiller<EthereumSigner>,
+    >,
+    RootProvider<RetryBackoffService<RuntimeTransport>, N>,
+    RetryBackoffService<RuntimeTransport>,
+    N,
+>;
+
+/// Helper type alias for a rpc url
+pub type RpcUrl = String;
```
Comment (mattsse) [inline crates/common/src/provider/mod.rs:44]: this alias is not useful

File: crates/common/src/provider/mod.rs [hunk 4]
```diff
@@ -1,7 +1,331 @@
...
+        let url = Url::parse(url_str)
+            .or_else(|err| match err {
+                ParseError::RelativeUrlWithoutBase => {
+                    if SocketAddr::from_str(url_str).is_ok() {
+                        Url::parse(&format!("http://{}", url_str))
+                    } else {
+                        let path = Path::new(url_str);
+
+                        if let Ok(path) = resolve_path(path) {
+                            Url::parse(&format!("file://{}", path.display()))
```
Comment (mattsse) [inline crates/common/src/provider/mod.rs:117]: this could be an issue for the windows ipc

**Review by mattsse — CHANGES_REQUESTED** [review]
nice,

a few more nits and suggestions

let's convert all for those

```rust
http_provider(&handle.http_endpoint());
```

to `handle.http_provider()`

same for ws and ipc

File: crates/anvil/tests/it/anvil_api.rs [hunk 5]
```diff
@@ -610,49 +603,51 @@ async fn test_fork_revert_call_latest_block_timestamp() {
...
 }
 
-#[tokio::test(flavor = "multi_thread")]
-async fn can_remove_pool_transactions() {
-    let (api, handle) = spawn(NodeConfig::test()).await;
-    let provider = ethers_http_provider(&handle.http_endpoint());
-    let wallet = handle.dev_wallets().next().unwrap().to_ethers();
-    let provider = Arc::new(SignerMiddleware::new(provider, wallet.clone()));
+// #[tokio::test(flavor = "multi_thread")]
+// async fn can_remove_pool_transactions() {
```
[the remark this answers is not in the document]
Comment (zerosnacks) [reply crates/anvil/tests/it/anvil_api.rs]: good catch, was previously blocked by missing txpool methods - now fixed

File: crates/anvil/tests/it/ipc.rs [hunk 6]
```diff
@@ -19,32 +19,39 @@ fn ipc_config() -> NodeConfig {
     NodeConfig::test().with_ipc(Some(Some(rand_ipc_endpoint())))
 }
 
+// TODO: throws: `Transport(Custom(Os { code: 2, kind: NotFound, message: "The system cannot find
+// the file specified." }))` on Windows
```
[the remark this answers is not in the document]
Comment (zerosnacks) [reply crates/anvil/tests/it/ipc.rs]: Got this error:

Transport(Custom(TransportErrorTest(Transport(Custom(Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." })), "file:///pipe/anvil-ipc-7861286456766970431", "\\\\.\\pipe\\anvil-ipc-7861286456766970431")))

The conversion appears to be correct (format: error, url, path), expected format is: `\\.\pipe\reth_engine_api.ipc` after escaping.

Attempted to replace it with the approach defined in this test in Alloy by using a tempfile: https://github.com/alloy-rs/alloy/blob/4d77029ba61284a6b4185f46581b4d9bce34f5fd/crates/rpc-client/tests/it/ipc.rs#L5

I noticed that Alloy currently only tests on latest Ubuntu making it likely that this is an area of the codebase not fully tested yet.

Going to hop onto my Windows machine and see if I can debug it there.

**Review by mattsse — APPROVED** [review]
I'm okay with disabling ipc tests on windows for now, and fix separately on alloy or in a follow up once we found it

but i really want more context (the path) included in this error, because this is useless:

> thread 'ipc::can_get_block_number_ipc' panicked at crates\anvil\tests\it\ipc.rs:31:49:
called `Result::unwrap()` on an `Err` value: Transport(Custom(TransportError(Transport(Custom(Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." })))))

this needs to happen in alloy though

**Comment by zerosnacks** [comment]
> I'm okay with disabling ipc tests on windows for now, and fix separately on alloy or in a follow up once we found it
> 
> but i really want more context (the path) included in this error, because this is useless:
> 
> > thread 'ipc::can_get_block_number_ipc' panicked at crates\anvil\tests\it\ipc.rs:31:49:
> > called `Result::unwrap()` on an `Err` value: Transport(Custom(TransportError(Transport(Custom(Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." })))))
> 
> this needs to happen in alloy though, do we know where this is thrown?

I agree, lets not make this a blocker and handle this in a follow up.

I've added some extra information to the `TransportError` in Foundry.

Some places I found suspicious:

casting to `ok()`

https://github.com/foundry-rs/foundry/blob/d21c3e7e1864d909cbece8ad650f91770193f128/crates/anvil/src/lib.rs#L324

Tests for the existence of the file, not something we do in the current test. However when I changed it to use Alloy's `NamedTempFile` of the example it didn't work.

https://github.com/alloy-rs/alloy/blob/bbef8de3fd203bf6feb26dc619f42b2bb2ecd4e3/crates/rpc-client/src/builtin.rs#L147-L162

And as previously highlighted

https://github.com/foundry-rs/foundry/blob/5f801b490d776e4f0e51a11c4d91de2c30e3ad3c/crates/common/src/provider/mod.rs#L113-L117

**Review by mattsse — APPROVED** [review]
gg

let's add a windows target to alloy CI and ensure IPC works cross platform

**Review by DaniPopes — APPROVED** [review]

**Comment by zerosnacks** [comment]
> gg
> 
> let's add a windows target to alloy CI and ensure IPC works cross platform

Opened a draft PR here: https://github.com/alloy-rs/alloy/pull/642

I previously made a mistake when I was testing it earlier against the CI (mistakingly had set `runs-on` to `ubuntu-latest`, not `{{ matrix.os }}`). The PR handles this correctly now and shows build failures / test failures for various reasons which I'll address in that PR.

**Review by klkvr — APPROVED** [review]

**Pull request merged**